File: 300-longest-increasing-subsequence/longest-increasing-subsequence.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        n=len(nums)
        cache=[1]*n
        LIS=-1
        for i in range(n):
            for j in range(i):
                if nums[i]>nums[j]:
                    cache[i]=max(cache[i],cache[j]+1)
            LIS=max(LIS,cache[i])
            
        return LIS

            

        

File: 300-longest-increasing-subsequence/longest-increasing-subsequence.cpp
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        vector<int> dp(nums.size(), 1);
        int res = 1;
        for (int i = 0; i < nums.size(); i++) {
            for (int j = 0; j < i; j++) 
                if (nums[j] < nums[i] && dp[j] + 1 > dp[i]) 
                    dp[i] = dp[j] + 1;
            res = max(dp[i], res);
        }
        return res;
    }
};